### PR TITLE
Fixing no focus border on the empty DGV in the Narrator

### LINKED ISSUE #2178: The empty DataGridView control cannot get a blue Narrator rectangle, just get the focus dot line
* .NET Core Version: (.NET Core SDK 3.1 Preview 2 3.1.100.14533)
* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**
The empty DataGridView control cannot get a blue Narrator rectangle, just get the focus dot line, so no any announcement about the empty dataGridView.
But when adding some columns to the DataGridView, the blue rectangle will on the currently specific cells with corresponding announcement by Narrator reading. Maybe this issue is by design, just report here and find the final confirmation from you guys.

This issue also can reproduce with latest .NET Framework 4.8 project.

**Actual behavior:** 
The empty DataGridView control cannot get a blue Narrator rectangle, just get the focus dot line.

**Expected behavior:**
The empty DataGridView control should get a blue Narrator rectangle.

**Minimal repro:**
1.	Run the attached app,
2.	Launch the Narrator and tab to navigate to the controls on the form.
3.	Observe the narrator blue rectangle between controls and DataGridView.
[WindowsFormsApp3.zip](https://github.com/dotnet/winforms/files/3766270/WindowsFormsApp3.zip)

### PR DESCRIPTION
**Pull request by DmitryGorokhov**
Fixes #2178  

## Proposed changes

- Reworked the property `HasKeyboardFocusPropertyId` of DGV AO. If no inner item then focus on the entire DGV else focus on the inner cell instead of entire DGV. This works also for DGV in Virtual mode. This doesn't break focusing border on the cell if DGV is not empty. It works also when form doesn't have any row (e.g. row for edit) see this case in after screens below.
- Turned off `TablePattern` support for DGV without rows. It needs to avoid incorrect announcement of the control:
if focus on empty DGV then it has no message "enter table", although if leave then it has message "exit table". So we turn off "enter table" and "exit table" messages for the empty DGV. This approach doesn't make a new accessibility problems in Accessibility Insights.
- Added unit tests for it.

## Customer Impact

- The Narrator has correct focus on empty DGV. 

## Regression? 

- No

## Risk

- Minimal

### Before

No a blue border focus on the empty DGV (only dashed border).

before

### After

The focusing blue border is set on the empty DGV.

after fix

If DGV is non editable e.g. DGV has `ReadOnly = true` and `AllowUserToAddRows = false` properties and DGV has no rows then the Narrator set the blue border on the entire DGV.

## Test methodology

- Manually
- Unit
- CTI

## Test environment(s) 

- Dotnet version 6.0.100
- OS Version 10.0.19043

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6102)

### CODE REVIEW

**Review by vladimir-krestov — APPROVED** [review]
Looks good ✔️ 
But I would make carefully testing different cases.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs [hunk 1]
```diff
@@ -239,7 +239,9 @@ internal override int[] RuntimeId
                     case UiaCore.UIA.NamePropertyId:
                         return Name;
                     case UiaCore.UIA.HasKeyboardFocusPropertyId:
-                        return false; // Only inner cell should be announced as focused by Narrator but not entire DGV.
+                        // If no inner cell entire DGV should be announced as focused by Narrator.
+                        // Else only inner cell should be announced as focused by Narrator but not entire DGV.
+                        return RowCount * ColumnCount == 0;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:244]: can we just say `return RowCount ==0`?

Comment (DmitryGorokhov) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs]: Yes, this is simpler and works the same, I checked it. So, let's use this approach. Thank you

**Review by dreddy-work — DISMISSED** [review]
LGTM.

**Comment by DmitryGorokhov** [comment]
Testers found an announcement problem. PR is still in progress.

**Comment by DmitryGorokhov** [comment]
Testers approved this changes and made sure that the regression has been fixed.

**Comment by Olina-Zhang** [comment]
No new issue found for the updated PR change.

**Review by dreddy-work — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs [hunk 2]
```diff
@@ -239,7 +239,9 @@ internal override int[] RuntimeId
                     case UiaCore.UIA.NamePropertyId:
                         return Name;
                     case UiaCore.UIA.HasKeyboardFocusPropertyId:
-                        return false; // Only inner cell should be announced as focused by Narrator but not entire DGV.
+                        // If no inner cell entire DGV should be announced as focused by Narrator.
+                        // Else only inner cell should be announced as focused by Narrator but not entire DGV.
+                        return RowCount == 0;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:244]: @vladimir-krestov  - should this be `return owner.Focused && RowCount == 0;`

**Pull request merged**